File: Java/384.shuffleAnArray.java
// class Solution {
//     private int[] array;
//     private int[] original;
//     private Random rand = new Random();
//     private List<Integer> getArrayCopy() {
//         List<Integer> asList = new ArrayList<Integer>();
//         for (int i = 0; i < array.length; i++) {
//             asList.add(array[i]);
//         }
//         return asList;
//     }
//     public Solution(int[] nums) {
//         array = nums;
//         original = nums.clone();
//     }
//     public int[] reset() {
//         array = original;
//         original = original.clone();
//         return array;
//     }
//     public int[] shuffle() {
//         List<Integer> aux = getArrayCopy();
//         for (int i = 0; i < array.length; i++) {
//             int removeIdx = rand.nextInt(aux.size());
//             array[i] = aux.get(removeIdx);
//             aux.remove(removeIdx);
//         }
//         return array;
//     }
// }

// 洗牌算法
class Solution {
    private int[] array;
    private int[] original;
    Random rand = new Random();
    private int randRange(int min, int max) {
        return rand.nextInt(max - min) + min;
    }
    private void swapAt(int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
    public Solution(int[] nums) {
        array = nums;
        original = nums.clone();
    }
    public int[] reset() {
        array = original;
        original = original.clone();
        return original;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        for (int i = 0; i < array.length; i++) {
            swapAt(i, randRange(i, array.length));
        }
        return array;
    }
}

File: Python/384.shuffleAnArray.py
# class Solution:
#     def __init__(self, nums: List[int]):
#         self.array = nums
#         self.original = list(nums)
#     def reset(self) -> List[int]:
#         self.array = self.original
#         self.original = list(self.original)
#         return self.array
#     def shuffle(self) -> List[int]:
#         aux = list(self.array)
#         for idx in range(len(self.array)):
#             remove_idx = random.randrange(len(aux))
#             self.array[idx] = aux.pop(remove_idx)
#         return self.array

# 洗牌算法
class Solution:
    def __init__(self, nums: List[int]):
        self.array = nums
        self.original = list(nums)
    def reset(self) -> List[int]:
        self.array = self.original
        self.original = list(self.original)
        return self.array
    def shuffle(self) -> List[int]:
        for i in range(len(self.array)):
            swap_idx = random.randrange(i, len(self.array))
            self.array[i], self.array[swap_idx] = self.array[swap_idx], self.array[i]
        return self.array